File: demos/python-feequote/scripts/release_notes.py
"""Print the commits between two git refs grouped by conventional-commit type.

Usage:
    python scripts/release_notes.py <from-ref> <to-ref>

Example:
    python scripts/release_notes.py v0.1.0 v0.2.0

Only the standard library is used and git is invoked through subprocess, so
the script runs unchanged on Windows, macOS and Linux. The output is a plain
grouped listing; turning it into the published release notes is a separate
editing step.
"""

from __future__ import annotations

import re
import subprocess
import sys
from collections import defaultdict
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

# Order matters: this is the order the groups are printed in.
GROUP_TITLES = {"feat": "Features", "fix": "Fixes", "chore": "Chores", "other": "Other"}

SUBJECT_PATTERN = re.compile(r"^(?P<type>[a-z]+)(?:\([^)]*\))?!?:\s*(?P<summary>.+)$")


def git_subjects(from_ref: str, to_ref: str) -> list[tuple[str, str]]:
    """Return (short hash, subject) for every non-merge commit in the range."""
    command = [
        "git",
        "log",
        "--no-merges",
        "--reverse",
        "--format=%h%x09%s",
        f"{from_ref}..{to_ref}",
    ]
    result = subprocess.run(
        command,
        cwd=REPO_ROOT,
        capture_output=True,
        text=True,
        encoding="utf-8",
        check=True,
    )
    subjects: list[tuple[str, str]] = []
    for line in result.stdout.splitlines():
        short_hash, _, subject = line.partition("\t")
        if short_hash:
            subjects.append((short_hash, subject.strip()))
    return subjects


def classify(subject: str) -> tuple[str, str]:
    """Split a commit subject into its group key and summary text."""
    match = SUBJECT_PATTERN.match(subject)
    if match is None:
        return "other", subject
    commit_type = match.group("type")
    group = commit_type if commit_type in GROUP_TITLES else "other"
    return group, match.group("summary")


def render(from_ref: str, to_ref: str, subjects: list[tuple[str, str]]) -> str:
    grouped: dict[str, list[str]] = defaultdict(list)
    for short_hash, subject in subjects:
        group, summary = classify(subject)
        grouped[group].append(f"- {summary} ({short_hash})")

    lines = [f"Commits {from_ref}..{to_ref}: {len(subjects)}", ""]
    for group, title in GROUP_TITLES.items():
        entries = grouped.get(group, [])
        if group == "other" and not entries:
            continue
        lines.append(f"{title} ({len(entries)})")
        lines.extend(entries or ["- none"])
        lines.append("")
    return "\n".join(lines).rstrip() + "\n"


def main(argv: list[str]) -> int:
    if len(argv) != 3:
        print("usage: python scripts/release_notes.py <from-ref> <to-ref>", file=sys.stderr)
        return 2
    from_ref, to_ref = argv[1], argv[2]
    try:
        subjects = git_subjects(from_ref, to_ref)
    except FileNotFoundError:
        print("error: git was not found on PATH", file=sys.stderr)
        return 1
    except subprocess.CalledProcessError as exc:
        reason = exc.stderr.strip().splitlines()[0] if exc.stderr.strip() else "unknown error"
        print(f"error: git log failed: {reason}", file=sys.stderr)
        return 1
    sys.stdout.write(render(from_ref, to_ref, subjects))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: demos/typescript-feequote/scripts/release-notes.mjs
#!/usr/bin/env node
// Prints the commits between two tags grouped by conventional-commit type.
// The release-notes skill turns this output into the house format.
//
//   node scripts/release-notes.mjs v0.1.0 v0.2.0
import { execFileSync } from 'node:child_process';

const GROUPS = [
  { key: 'feature', title: 'feature', types: ['feat'] },
  { key: 'fix', title: 'fix', types: ['fix'] },
  { key: 'chore', title: 'chore', types: ['chore', 'build', 'ci', 'docs', 'refactor', 'style', 'test', 'perf'] },
];

const SUBJECT_PATTERN = /^(?<type>[a-z]+)(?:\([^)]*\))?!?:\s*(?<summary>.+)$/;

function usage() {
  console.error('Usage: node scripts/release-notes.mjs <from-tag> <to-tag>');
  process.exit(2);
}

function readSubjects(fromRef, toRef) {
  const output = execFileSync('git', ['log', '--no-merges', '--format=%s', `${fromRef}..${toRef}`], {
    encoding: 'utf8',
    stdio: ['ignore', 'pipe', 'inherit'],
  });
  return output
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
}

function groupSubjects(subjects) {
  const grouped = new Map(GROUPS.map((group) => [group.key, []]));
  const other = [];
  for (const subject of subjects) {
    const match = SUBJECT_PATTERN.exec(subject);
    const group = match ? GROUPS.find((candidate) => candidate.types.includes(match.groups.type)) : undefined;
    if (match && group) {
      grouped.get(group.key).push(match.groups.summary);
    } else {
      other.push(subject);
    }
  }
  return { grouped, other };
}

function printGroup(title, entries) {
  console.log(`${title}:`);
  if (entries.length === 0) {
    console.log('  (none)');
  }
  for (const entry of entries) {
    console.log(`  - ${entry}`);
  }
}

const [fromRef, toRef] = process.argv.slice(2);
if (!fromRef || !toRef) {
  usage();
}

let subjects;
try {
  subjects = readSubjects(fromRef, toRef);
} catch (error) {
  console.error(`Could not read the git log for ${fromRef}..${toRef}: ${error.message}`);
  process.exit(1);
}

const { grouped, other } = groupSubjects(subjects);
console.log(`Commits ${fromRef}..${toRef} (${subjects.length} total)`);
for (const group of GROUPS) {
  printGroup(group.title, grouped.get(group.key));
}
if (other.length > 0) {
  printGroup('other (not conventional)', other);
}
